File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0988_Smallest_String_Starting_From_Leaf.java
package com.lanceolata.coding.algorithms.java.leetcode;

public class Question_0988_Smallest_String_Starting_From_Leaf {
    // Definition for a binary tree node.
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    private String res;

    public String smallestFromLeaf(TreeNode root) {
        res = null;
        dfs(root, new StringBuilder());
        return res;
    }

    private void dfs(TreeNode node, StringBuilder sb) {
        if (node == null) {
            return;
        }
        sb.append((char) ('a' + node.val));
        if (node.left == null && node.right == null) {
            sb.reverse();
            String cur = sb.toString();
            sb.reverse();
            if (res == null || cur.compareTo(res) < 0) {
                res = cur;
            }
        }
        dfs(node.left, sb);
        dfs(node.right, sb);
        sb.deleteCharAt(sb.length() - 1);
    }
}

File: algorithms/algorithms-cplusplus/leetcode/Question_0988_Smallest_String_Starting_From_Leaf.cc
#include <string>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
 public:
  string smallestFromLeaf(TreeNode* root) {
    string res;
    dfs(root, "", res);
    return res;
  }

 private:
  void dfs(TreeNode* root, string s, string& res) {
    if (!root) {
      return;
    }
    s = char(root->val + 'a') + s;
    if (!root->left && !root->right){
      res = res == "" ? s : min(res, s);
    }
    dfs(root->left, s, res);
    dfs(root->right, s, res);
  }
};
